### ISSUE TITLE
Tree generation broken for multiple roots

### ISSUE THREAD

**Issue by jeromekelleher** [jeromekelleher/msprime#235]
When we have multiple roots in trees we can now get an error from recently added error checking code. For example:
```
id      flags   population      time
0       1       -1              0.00000000000000
1       1       -1              0.00000000000000
2       1       -1              0.00000000000000
3       1       -1              0.00000000000000
4       1       -1              0.00000000000000
5       1       -1              0.00000000000000
6       1       -1              0.00000000000000
7       1       -1              0.00000000000000
8       1       -1              0.00000000000000
9       1       -1              0.00000000000000
10      0       -1              1.00000000000000
11      0       -1              1.00000000000000
12      0       -1              1.00000000000000
13      0       -1              1.00000000000000
id      left            right           parent  children
0       0.00000000      0.93423354      10      0,3
1       0.00000000      0.57828326      11      1,9
2       0.00000000      1.00000000      12      5,8
3       0.93423354      1.00000000      13      0,7
```

When running the trees iterator over this we get 
```
LibraryError: Bad edgesets: contradictory children for a given parent over an interval.
```
This is one of a number of issues to tackle with multiple roots.

**Comment by jeromekelleher** [comment]
Keeping track of the number of roots seems easy enough and efficient to do, but keeping track of the actual set of roots seems more difficult to me. I'm not sure how to do this without incurring a log(#num nodes) cost for each edgeset we insert and remove.

The first question we have to tackle is what's the definition of a root. The most useful definition of a root I think is a 'sample root'. That is, the roots of a tree are the set of nodes that terminate the paths from all samples. However, there are other definitions which might be worth thinking about.

To count the number of sample roots we do the following. Add a new binary map to the ``sparse_tree_t`` called `is_root`. Initially, set ``is_root[u] = 1`` for all samples ``u``. Tree transitions work as follows. For each edgeset we remove, set ``is_root[c] = is_root[parent]`` for each child ``c``. For each edgeset we insert: if ``is_root[c] == 1``, set ``is_root[parent] = 1`` and set ``is_root[c] = 0`` for each child c.

This way, we can keep track of the roots in the map ``is_root``, and it's simple enough to see how we can count the roots. I don't see how we can efficiently keep track of what the roots actually are though without using either an AVL tree  (incurring a O(log # samples) + memory management overhad) or a hashtable (incurring a hefty extra constant) to store the current set of roots. 

If this was a common operation this would seem like a fair cost, but I would imagine that the vast majority of tree sequences will have single rooted trees, where this overhead is significant and a complete waste of time.

**Comment by jeromekelleher** [comment]
The skipped test in ``test_wright_fisher.py`` is an instance of this problem:

https://github.com/jeromekelleher/msprime/blob/master/tests/test_wright_fisher.py#L198

**Comment by jeromekelleher** [comment]
Actually, thinking about this a bit more, I think we can solve the problem efficiently with a circular linked list of roots. Dealing with multiple roots properly is a precondition for thoroughly testing the WF simulations (and simplify) so I'm going to see if I can implement it.

**Comment by jeromekelleher** [comment]
Fixed in #269.
